content/blog/buildkite-cd.md

```diff
@@ -0,0 +1,136 @@
+
+```
+  - wait
+
+  # This makes sure that deploys are triggered in the same order as the
+  # test builds, no matter which test builds finish first.
+  # see https://buildkite.com/docs/pipelines/controlling-concurrency
+  - label: "Concurrency gate"
+    command: "exit 0"
+    if: build.branch == 'main'
+    concurrency: 1
+    concurrency_group: "app-concurrency-gate"
+
+  - wait
+
+  - label: ":rocket:"
+    trigger: "deploy"
+    if: build.branch == 'main'
+```
+
+Buildkite can support more complex [triggers](https://buildkite.com/docs/pipelines/trigger-step) and [scheduled builds](https://buildkite.com/docs/pipelines/scheduled-builds). For our use case, we have a pipeline that deploys to a staging instance which is triggered automatically whenever there is a change to any of the services which form the app.
+
+
+# Controlling deployment
+
+We would also like to have the option to deploy a specific branch of one of our services so we can see the effect of big changes before merging into main. Buildkite gives us a way to do this through [environent variables](https://buildkite.com/docs/pipelines/environment-variables#defining-your-own). Each service that forms part of the app has a CI pipeline that builds a docker image. These are tagged with the branch name and the sha. We can use environment variables to set the tag of the docker image we want to deploy. There are multiple ways to set environment variables but in this example we set them through the pipeline yml. Update your deployment pipeline to look like
+
+```
+env:
+  TAG: "${TAG:-main}"
+
+steps:
+  - label: ":rocket: deploy"
+    command: >-
+      ssh -i ~/.ssh/id_deploy -oStrictHostKeyChecking=accept-new <username>@<host> './deploy.sh --tag=$TAG'
+    agents:
+      queue: "deploy"
+```
+
+This will define an env var `TAG` with default value `main` which will then be passed into the command step. The deployment script will have to take a `--tag` argument to bring up the docker container with that specific tag.
+
+When we run a build via the Buildkite UI we can then set the environment variable in the new build dialog.
+
+<img src="/img/buildkite-cd-run.png" alt="png of pipeline run"/>
+
+# Next steps
+
+We wanted to set up Buildkite deployment pipelines to both simplify deployments and automate repeated manual work. We now have pipelines that will allow us to deploy manually, deploy automatically, and deploy a specified tag of a docker image. When we deploy to the staging instance we still have to manually check the newly deployed version works. This is slow and prone to missing regressions. When we deploy a new feature we manually check it, but we don't necessarily check all other parts of the app. Bugs can be introduced which we only know about when users report issues or we notice errors in the logs. We could take the automation a step further by writing browser tests that run after the deployment.
```

Comment:
This paragraph sort of just trails off and you might turn it into something more positive and active - say we are planning on automatically running browser tests against the deployed instances, but then say something good about what the system in its current state helps us (you already have this information in the introduction but the reader has forgotten it by now)